### ISSUE TITLE
Predicted output is black.

### ISSUE THREAD

**Issue by akshay9396** [milesial/Pytorch-UNet#506]
I trained the model on Carvana dataset. Below is the screen shot of training.  There i you can clearly see that validation Dice score is constant for 5 epochs and loss is "nan". I used one checkpoint.pth for predication and i got black output.  could you please help me to resolved this issue.
![image](https://github.com/user-attachments/assets/082e28e9-7389-4bdc-95c8-b6f956d0fee1)
![Capture](https://github.com/user-attachments/assets/886a1bc9-d10d-475f-aaa9-f168f9a52972)

**Comment by Feng2100** [comment]
Have you solved it? I'm having the exact same problem

**Comment by chenshans** [comment]
I'm having the exact same problem，too

**Comment by berda-ak** [comment]
I'm having the exact same problem，too

**Comment by caigg188** [comment]
I'm having the exact same problem，too

**Comment by taipain** [comment]
> I trained the model on Carvana dataset. Below is the screen shot of training. There i you can clearly see that validation Dice score is constant for 5 epochs and loss is "nan". I used one checkpoint.pth for predication and i got black output. could you please help me to resolved this issue. ![image](https://private-user-images.githubusercontent.com/109768122/354624121-082e28e9-7389-4bdc-95c8-b6f956d0fee1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1NTI5MzAsIm5iZiI6MTczMTU1MjYzMCwicGF0aCI6Ii8xMDk3NjgxMjIvMzU0NjI0MTIxLTA4MmUyOGU5LTczODktNGJkYy05NWM4LWI2Zjk1NmQwZmVlMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExNFQwMjUwMzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYzEwNmIwYWNjN2YzYjc3NDgxN2IxMTc1M2NmYjMxNmMyZWE0YjVlMzA0MTA1ODYxMDY4NmE5Yzk4ODc4MmY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.OxsSJ_SSTQZaAKSPFTZtRygfzZDsPRU4FdJmwSI-az4) ![Capture](https://private-user-images.githubusercontent.com/109768122/354624292-886a1bc9-d10d-475f-aaa9-f168f9a52972.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1NTI5MzAsIm5iZiI6MTczMTU1MjYzMCwicGF0aCI6Ii8xMDk3NjgxMjIvMzU0NjI0MjkyLTg4NmExYmM5LWQxMGQtNDc1Zi1hYWE5LWYxNjhmOWE1Mjk3Mi5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExNFQwMjUwMzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hYjZhYzEzNzA0M2ZkNWY1ZTQ3YjBlOGUzMDdmOWNjMmIzNGQ3MjA1MGUwOTE2Yzc1ZDJmNDgwMzMxNWFjZDI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.A_hYXDPt3YYw7DX6YEAu64boNrkVP5JMsHODdO10RDM)

Probably the reason is that torch.float16 is used in autocast() when using AMP.
So, change torch.float16 to torch.bfloat16. 
But, if you wanna use torch.float16, you add grad_scaler in init_scale=4096．
I solved this issue.
I referred to the following sites．
This is my first time commenting on github, so sorry if there is something wrong and  my poor English．
https://qiita.com/takeuchiseijin/items/909c48b57127a37fbd12
https://qiita.com/bowdbeg/items/71c62cf8ef891d164ecd

**Comment by MorleyOlsen** [comment]
> > I trained the model on Carvana dataset. Below is the screen shot of training. There i you can clearly see that validation Dice score is constant for 5 epochs and loss is "nan". I used one checkpoint.pth for predication and i got black output. could you please help me to resolved this issue. ![image](https://private-user-images.githubusercontent.com/109768122/354624121-082e28e9-7389-4bdc-95c8-b6f956d0fee1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1NTI5MzAsIm5iZiI6MTczMTU1MjYzMCwicGF0aCI6Ii8xMDk3NjgxMjIvMzU0NjI0MTIxLTA4MmUyOGU5LTczODktNGJkYy05NWM4LWI2Zjk1NmQwZmVlMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExNFQwMjUwMzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYzEwNmIwYWNjN2YzYjc3NDgxN2IxMTc1M2NmYjMxNmMyZWE0YjVlMzA0MTA1ODYxMDY4NmE5Yzk4ODc4MmY0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.OxsSJ_SSTQZaAKSPFTZtRygfzZDsPRU4FdJmwSI-az4) ![Capture](https://private-user-images.githubusercontent.com/109768122/354624292-886a1bc9-d10d-475f-aaa9-f168f9a52972.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE1NTI5MzAsIm5iZiI6MTczMTU1MjYzMCwicGF0aCI6Ii8xMDk3NjgxMjIvMzU0NjI0MjkyLTg4NmExYmM5LWQxMGQtNDc1Zi1hYWE5LWYxNjhmOWE1Mjk3Mi5QTkc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExNFQwMjUwMzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hYjZhYzEzNzA0M2ZkNWY1ZTQ3YjBlOGUzMDdmOWNjMmIzNGQ3MjA1MGUwOTE2Yzc1ZDJmNDgwMzMxNWFjZDI1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.A_hYXDPt3YYw7DX6YEAu64boNrkVP5JMsHODdO10RDM)
> 
> Probably the reason is that torch.float16 is used in autocast() when using AMP. So, change torch.float16 to torch.bfloat16. But, if you wanna use torch.float16, you add grad_scaler in init_scale=4096． I solved this issue. I referred to the following sites． This is my first time commenting on github, so sorry if there is something wrong and my poor English． https://qiita.com/takeuchiseijin/items/909c48b57127a37fbd12 https://qiita.com/bowdbeg/items/71c62cf8ef891d164ecd

so to solve this problem, we simply should add dtype in autocast, which appears as 'with torch.autocast(dtype = torch.bfloat16)' in the train.py. (at least that is what i have thought from your links, qwq)

**Comment by RalphFH** [comment]
@taipain thanks for the solution. It helps to set ' dtype=torch.bfloat16' in meth:torch.autocast, which address the problem of 'NAN' loss and black predition.

**Comment by algwang** [comment]
> 感谢您的解决方案。在 meth：torch.autocast 中设置 'dtype=torch.bfloat16' 有助于解决 'NAN' 丢失和黑色 predition 的问题。
Hi, could you please tell me which folder it is in?

**Comment by taipain** [comment]
> > 感谢您的解决方案。在 meth：torch.autocast 中设置 'dtype=torch.bfloat16' 有助于解决 'NAN' 丢失和黑色 predition 的问题。
> Hi, could you please tell me which folder it is in?

train.py
before
with torch.autocast(device.type if device.type != 'mps' else 'cpu', enabled=amp):

after
with torch.autocast(device.type if device.type != 'mps' else 'cpu', enabled=amp,dtype=torch.bfloat16):

**Comment by algwang** [comment]
> > > 感谢您的解决方案。在 meth：torch.autocast 中设置 'dtype=torch.bfloat16' 有助于解决 'NAN' 丢失和黑色 predition 的问题。您好，您能告诉我它在哪个文件夹中吗？
> 
> 之前使用 torch.autocast（device.type if device.type ！= 'mps' else 'cpu'， enabled=amp）train.py：
> 
> 之后使用 torch.autocast（device.type if device.type ！= 'mps' else 'cpu'， enabled=amp，dtype=torch.bfloat16）：

Hello, after I made the modification, I still get the same error. When I added the validation set during training, the images of the validation machine were masked, but when I used predict.py for testing, I still got the same error. Can you explain why?
